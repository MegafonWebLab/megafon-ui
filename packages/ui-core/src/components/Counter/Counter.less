@import '~styles/base';
@b: e('.mfui-beta-counter');

@text-color-disabled: rgba(@freshAsphalt, 0.5);

.disabled() {
    cursor: default;
    background-color: @spbSky1;
}

.iconDisabled() {
    fill: @text-color-disabled;
}

@{b} {
    background-color: @clearWhite;
    box-sizing: border-box;
    display: flex;
    justify-content: space-between;
    height: 32px;
    width: 120px;

    &__btn {
        background-color: transparent;
        border: 1px solid @spbSky2;
        box-sizing: border-box;
        cursor: pointer;
        display: flex;
        height: 100%;
        font-size: 20px;
        flex-grow: 1;
        align-items: center;
        justify-content: center;
        margin: 0;
        outline: none;
        padding: 0;
        appearance: none;
        -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
        text-decoration: none;
        max-width: 32px;
        width: 100%;
        flex-shrink: 0;

        &:hover {
            background-color: @spbSky0;
        }

        &:disabled {
            .disabled();

            @{b}__icon {
                .iconDisabled();
            }
        }
    }

    &__btn_left {
        border-top-left-radius: 50px;
        border-bottom-left-radius: 50px;
    }

    &__btn_right {
        border-top-right-radius: 50px;
        border-bottom-right-radius: 50px;
    }

    &__icon {
        display: block;
        height: 20px;
        width: 20px;
    }

    &__input-box {
        border-top: 1px solid @spbSky2;
        border-bottom: 1px solid @spbSky2;
        box-sizing: border-box;
        display: flex;
        align-items: center;
        overflow: hidden;
        height: 100%;
    }

    &__input {
        border: none;
        box-sizing: border-box;
        cursor: text;
        height: 100%;
        appearance: none;
        box-shadow: none;
        overflow: auto;
        outline: none;
        margin: 0;
        padding: 0;
        text-align: center;
        width: 100%;
    }

    &_disabled {
        @{b}__btn {
            .disabled();
        }

        @{b}__input {
            .disabled();
            color: @text-color-disabled;
        }

        @{b}__icon {
            .iconDisabled();
        }
    }
}
