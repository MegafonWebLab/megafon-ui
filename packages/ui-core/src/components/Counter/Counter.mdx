---
name: Counter
menu: Компоненты
route: components/counter
---

import { Playground, Props } from 'docz';
import Counter from './Counter';
import { DemoCounterWrapper, flexStyle } from './Counter.docz';

# Counter

## Properties

<Props of={Counter} />

```javascript
export const DemoCounterWrapper: React.FC<IDemoCounterWrapperProps> = ({ initialValue = 0, children }) => {
    const [value, setValue] = React.useState(initialValue);

    return (
        <div>
            <p style={{ marginTop: 0 }}>Value: {value}</p>
            {children({
                onChange: setValue,
                initialValue: initialValue,
            })}
        </div>
    );
};
```

## Basic usage

<Playground>
    <div style={flexStyle}>
        <DemoCounterWrapper initialValue={3}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper initialValue={18}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper initialValue={33}>
            {({ onChange, initialValue }) =>
                <Counter
                    initialValue={initialValue}
                    onChange={onChange}
                    min={3}
                    max={33}
                />
            }
        </DemoCounterWrapper>
        <DemoCounterWrapper>
            {({ onChange }) =>
                <Counter
                    onChange={onChange}
                    isDisabled
                />
            }
        </DemoCounterWrapper>
    </div>
</Playground>
