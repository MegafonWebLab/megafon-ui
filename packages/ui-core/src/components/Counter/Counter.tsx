import * as React from 'react';
import * as PropTypes from 'prop-types';
import cnCreate from 'utils/cnCreate';
import './Counter.less';
import IconMinus from 'icons/System/16/Minus_16.svg';
import IconPlus from 'icons/System/16/Plus_16.svg';

export interface ICounterProps {
    /** Начальное значение */
    initialValue?: number;
    /** Минимальное доступное значение */
    min?: number;
    /** Максимальное доступное значение */
    max?: number;
    /** Запретить взаимодействие с компонентом */
    isDisabled?: boolean;
    /** Обработчик изменения значения 'value' */
    onChange?: (value: number) => void;
    /** Дополнительный класс для корневого элемента */
    className?: string;
    /** Дополнительные классы для корневого и внутренних элементов */
    classes?: {
        root?: string;
        buttonMinus?: string;
        buttonPlus?: string;
        input?: string;
    };
}

const cn = cnCreate('mfui-beta-counter');
const Counter: React.FC<ICounterProps> = ({
    initialValue,
    max = 999999,
    min = 0,
    isDisabled = false,
    onChange,
    className,
    classes = {},
}) => {
    const currentInitialValue = initialValue || min;
    const [counter, setCounter] = React.useState(currentInitialValue);

    React.useEffect(() => {
        setCounter(currentInitialValue);
    }, [currentInitialValue]);

    const handleValueChange = React.useCallback((value: number) => {
        setCounter(value);

        if (value < min) {
            onChange && onChange(min);
        }

        if (value > max) {
            onChange && onChange(max);
        }

        if (value >= min && value <= max) {
            onChange && onChange(value);
        }
    }, [min, max, onChange]);

    const handleMinusClick = React.useCallback((): void => {
        handleValueChange(counter - 1);
    }, [handleValueChange, counter]);

    const handlePlusClick = React.useCallback((): void => {
        handleValueChange(counter + 1);
    }, [handleValueChange, counter]);

    const handleInputChange = React.useCallback((e: React.ChangeEvent<HTMLInputElement>): void => {
        const pattern = /^[0-9\b]+$/;
        const { value } = e.target;

        if (value !== '' && !pattern.test(value)) {
            return;
        }

        handleValueChange(Number(value));
    }, [handleValueChange]);

    const handleInputBlur = React.useCallback((e: React.FocusEvent<HTMLInputElement>): void => {
        const { value } = e.target;
        const numberValue = Number(value);

        if (numberValue < min) {
            handleValueChange(min);

            return;
        }

        if (numberValue > max) {
            handleValueChange(max);

            return;
        }
    }, [handleValueChange, min, max]);

    return (
        <div className={cn({ disabled: isDisabled }, [className, classes.root])}>
            <button
                className={cn('btn', { left: true }, classes.buttonMinus)}
                type="button"
                disabled={isDisabled || counter <= min}
                onClick={handleMinusClick}
            >
                <IconMinus className={cn('icon')} />
            </button>
            <div className={cn('input-box')}>
                <input
                    className={cn('input', classes.input)}
                    value={counter}
                    onChange={handleInputChange}
                    onBlur={handleInputBlur}
                    disabled={isDisabled}
                />
            </div>
            <button
                className={cn('btn', { right: true }, classes.buttonPlus)}
                type="button"
                disabled={isDisabled || counter >= max}
                onClick={handlePlusClick}
            >
                <IconPlus className={cn('icon')} />
            </button>
        </div>
    );
};

Counter.propTypes = {
    initialValue: PropTypes.number,
    min: PropTypes.number,
    max: PropTypes.number,
    isDisabled: PropTypes.bool,
    onChange: PropTypes.func,
    className: PropTypes.string,
    classes: PropTypes.shape({
        root: PropTypes.string,
        buttonMinus: PropTypes.string,
        buttonPlus: PropTypes.string,
        input: PropTypes.string,
    }),
};

export default Counter;
