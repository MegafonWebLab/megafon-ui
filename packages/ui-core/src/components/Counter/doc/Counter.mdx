---
name: Counter
menu: Компоненты
route: components/counter
---

import { Playground, Props } from 'docz';
import Counter from '../Counter';
import Example from './Counter.example.mdx';

# Counter

<CopyToClipboardBox text="import { Counter } from '@megafon/ui-core';" />

<DoczTabs>
    <Example />
    <Props of={Counter} />
</DoczTabs>
