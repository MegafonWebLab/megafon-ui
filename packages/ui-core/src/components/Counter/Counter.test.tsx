import * as React from 'react';
import { shallow, mount } from 'enzyme';
import Counter, { ICounterProps } from './Counter';

const props: ICounterProps = {
    initialValue: 10,
    min: 3,
    max: 33,
    isDisabled: true,
    className: 'class-name',
    classes: {
        root: 'root-class',
        buttonMinus: 'button-minus-class',
        buttonPlus: 'button-plus-class',
        input: 'input-class',
    },
};

describe('<Counter />', () => {
    describe('layout', () => {
        it('renders Counter', () => {
            const wrapper = shallow(
                <Counter />
            );
            expect(wrapper).toMatchSnapshot();
        });

        it('renders Counter with props', () => {
            const wrapper = shallow(
                <Counter {...props} />
            );
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('handlers', () => {
        it('calls onChange for buttons', () => {
            const handleClick = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={10} onChange={handleClick} />
            );

            wrapper.find('button').first().simulate('click');
            expect(handleClick).toHaveBeenCalledWith(9);

            wrapper.find('button').at(1).simulate('click');
            expect(handleClick).toHaveBeenCalledWith(10);
        });

        it('calls onChange for input', () => {
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Counter onChange={handleChange} />
            );

            wrapper.find('input').simulate('change', { target: { value: 22 } });
            expect(handleChange).toHaveBeenCalledWith(22);
        });

        it('calls onChange for input with wrong value', () => {
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Counter initialValue={10} onChange={handleChange} />
            );

            wrapper.find('input').simulate('change', { target: { value: 'hello' } });
            expect(handleChange).not.toBeCalled();
        });

        it('calls onChange for input with value less than min limit and more than max limit', () => {
            const handleChange = jest.fn();

            const wrapper = shallow(
                <Counter min={3} max={33} onChange={handleChange} />
            );

            wrapper.find('input').simulate('change', { target: { value: 2 } });
            expect(handleChange).toHaveBeenCalledWith(3);

            wrapper.find('input').simulate('change', { target: { value: 34 } });
            expect(handleChange).toHaveBeenCalledWith(33);
        });

        it('calls onChange for left button near min limit', () => {
            const handleClick = jest.fn();

            const wrapper = mount(
                <Counter initialValue={3} min={3} onChange={handleClick} />
            );

            wrapper.find('button').first().simulate('click');
            expect(handleClick).not.toHaveBeenCalled();
        });

        it('calls onChange for right button near max limit', () => {
            const handleClick = jest.fn();

            const wrapper = mount(
                <Counter initialValue={33} max={33} onChange={handleClick} />
            );

            wrapper.find('button').at(1).simulate('click');
            expect(handleClick).not.toHaveBeenCalled();
        });
    });
});
