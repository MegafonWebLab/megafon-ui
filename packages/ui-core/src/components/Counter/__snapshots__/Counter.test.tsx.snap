// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> layout renders Counter 1`] = `
<div
  className="mfui-beta-counter"
>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_left"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <test-file-stub
      className="mfui-beta-counter__icon"
    />
  </button>
  <div
    className="mfui-beta-counter__input-box"
  >
    <input
      className="mfui-beta-counter__input"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      value={0}
    />
  </div>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_right"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <test-file-stub
      className="mfui-beta-counter__icon"
    />
  </button>
</div>
`;

exports[`<Counter /> layout renders Counter with props 1`] = `
<div
  className="mfui-beta-counter mfui-beta-counter_disabled class-name root-class"
>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_left button-minus-class"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <test-file-stub
      className="mfui-beta-counter__icon"
    />
  </button>
  <div
    className="mfui-beta-counter__input-box"
  >
    <input
      className="mfui-beta-counter__input input-class"
      disabled={true}
      onBlur={[Function]}
      onChange={[Function]}
      value={10}
    />
  </div>
  <button
    className="mfui-beta-counter__btn mfui-beta-counter__btn_right button-plus-class"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    <test-file-stub
      className="mfui-beta-counter__icon"
    />
  </button>
</div>
`;
